src/utils/contracts.ts

```diff
@@ -1,42 +1,82 @@
+  INTERFACE_NOT_SUPPORTED = "INTERFACE_NOT_SUPPORTED",
+  ORDER_NOT_VALID = "ORDER_NOT_VALID",
+  POLL_TRY_NEXT_BLOCK = "POLL_TRY_NEXT_BLOCK",
+  POLL_NEVER = "POLL_NEVER",
+  POLL_TRY_AT_BLOCK = "POLL_TRY_AT_BLOCK",
+  POLL_TRY_AT_EPOCH = "POLL_TRY_AT_EPOCH",
+}
+
+type ParsedCustomError = {
+  [K in CustomErrorSelectors]: K extends
+    | CustomErrorSelectors.PROOF_NOT_AUTHED
+    | CustomErrorSelectors.SINGLE_ORDER_NOT_AUTHED
+    | CustomErrorSelectors.INTERFACE_NOT_SUPPORTED
+    | CustomErrorSelectors.INVALID_FALLBACK_HANDLER
+    | CustomErrorSelectors.INVALID_HANDLER
+    | CustomErrorSelectors.SWAP_GUARD_RESTRICTED
+    ? { selector: K }
+    : K extends
+        | CustomErrorSelectors.ORDER_NOT_VALID
+        | CustomErrorSelectors.POLL_TRY_NEXT_BLOCK
+        | CustomErrorSelectors.POLL_NEVER
+    ? { selector: K; message: string }
+    : K extends
+        | CustomErrorSelectors.POLL_TRY_AT_BLOCK
+        | CustomErrorSelectors.POLL_TRY_AT_EPOCH
+    ? { selector: K; message: string; blockNumberOrEpoch: number }
+    : never;
+}[CustomErrorSelectors];
+
+export const CUSTOM_ERROR_ABI_MAP: Record<CustomErrorSelectors, string> = {
+  [CustomErrorSelectors.PROOF_NOT_AUTHED]: "ProofNotAuthed()",
+  [CustomErrorSelectors.SINGLE_ORDER_NOT_AUTHED]: "SingleOrderNotAuthed()",
+  [CustomErrorSelectors.SWAP_GUARD_RESTRICTED]: "SwapGuardRestricted()",
+  [CustomErrorSelectors.INVALID_HANDLER]: "InvalidHandler()",
+  [CustomErrorSelectors.INVALID_FALLBACK_HANDLER]: "InvalidFallbackHandler()",
+  [CustomErrorSelectors.INTERFACE_NOT_SUPPORTED]: "InterfaceNotSupported()",
+  [CustomErrorSelectors.ORDER_NOT_VALID]: "OrderNotValid(string)",
+  [CustomErrorSelectors.POLL_TRY_NEXT_BLOCK]: "PollTryNextBlock(string)",
+  [CustomErrorSelectors.POLL_NEVER]: "PollNever(string)",
+  [CustomErrorSelectors.POLL_TRY_AT_BLOCK]: "PollTryAtBlock(uint256,string)",
+  [CustomErrorSelectors.POLL_TRY_AT_EPOCH]: "PollTryAtEpoch(uint256,string)",
 };
 
-// Just export the keys of the CUSTOM_ERROR_ABI_MAP as a type
-export type CustomError = keyof typeof CUSTOM_ERROR_ABI_MAP;
+// Process the CUSTOM_ERROR_ABI_MAP to change the values to the ABI-encoded selectors
+const CUSTOM_ERROR_SELECTOR_MAP: Record<string, CustomErrorSelectors> = {};
 
-// To be able to do fast lookup of the selector given the id, we need to reverse the map
-// This should be a one-time operation, and cached for future use
-const CUSTOM_ERROR_ID_TO_NAME_MAP = Object.entries(CUSTOM_ERROR_ABI_MAP).reduce(
-  (acc, [name, selector]) => {
-    acc[ethers.utils.id(selector).slice(0, 10)] = name as CustomError;
-    return acc;
-  },
-  {} as Record<string, CustomError>
-);
+export const abiToSelector = (abi: string): string =>
+  ethers.utils.id(abi).slice(0, 10);
+
+for (const errorType of Object.keys(
+  CUSTOM_ERROR_ABI_MAP
+) as CustomErrorSelectors[]) {
+  const selector = abiToSelector(CUSTOM_ERROR_ABI_MAP[errorType]);
+  CUSTOM_ERROR_SELECTOR_MAP[selector] = errorType;
+}
```

Comment:
I guess this is a init code. it wiuld be more clear if you do a IIFE to asign the inital value to `CUSTOM_ERROR_SELECTOR_MAP` or you create a named function that gets the value and you assign it to it.

Reply:
Done.